engine/src/witness/sol/program_swaps_witnessing.rs

```diff
@@ -0,0 +1,544 @@
+											source_chain: cf_primitives::ForeignChain::Solana, // TODO: Pass chain id from above?
+											source_address: Some(ForeignChainAddress::Sol(data.sender.into())),
+											channel_metadata: CcmChannelMetadata {
+												message: ccm_parameters.message
+													.to_vec()
+													.try_into()
+													.map_err(|_| anyhow!("Failed to deposit CCM: `message` too long.")).ok()?,
+												gas_budget: ccm_parameters.gas_amount.into(),
+												ccm_additional_data,
+											},
+										});
+										(deposit_metadata, vault_swap_parameters)
+									}
+								};
+
+								Some(Ok((VaultSwapAccountAndSender {
+									vault_swap_account: account,
+									swap_sender: data.sender.into()
+								}, SolanaVaultSwapDetails {
+									from: if data.src_token.is_none() {SolAsset::Sol} else {SolAsset::SolUsdc},
+									deposit_amount: data.amount,
+									destination_address: EncodedAddress::from_chain_bytes(data.dst_chain.try_into().map_err(|e| warn!("error while parsing destination chain for solana vault swap:{}. Omitting swap", e)).ok()?, data.dst_address.to_vec()).map_err(|e| warn!("failed to decode the destination chain address for solana vault swap:{}. Omitting swap", e)).ok()?,
+									to: data.dst_token.try_into().map_err(|e| warn!("error while decoding destination token for solana vault swap: {}. Omitting swap", e)).ok()?,
+									deposit_metadata,
+									// TODO: These two will potentially be a TransactionId type
+									swap_account: account,
+									creation_slot: data.creation_slot,
+									broker_fees: vault_swap_parameters.broker_fees,
+									refund_params: Some(vault_swap_parameters.refund_params),
+									dca_params: vault_swap_parameters.dca_params,
+									boost_fee: vault_swap_parameters.boost_fee,
+								})))
+							}
+
+					// It could happen that some account is closed between the queries. This should
+					// not happen because:
+					// 1. Accounts in `new_program_swap_accounts` can only be accounts that have
+					//    newly been opened and they won't be closed until consensus is reached.
+					// 2. If due to rpc load management the get event accounts rpc is queried at a
+					//    slot < get swap endpoint data rpc slot, the min_context_slot will prevent
+					//    it from being executed before that.
+					// This could only happen if an engine is behind and were to see the account
+					// opened and closed between queries. That's not reallistic as it takes minutes
+					// for an account to be closed and even if it were to happen it's not
+					// problematic as we'd have reached consensus and the engine would just filter
+					// it out.
+					None => {
+						warn!("Event account not found for solana event account");
+						None
+					},
+					_ => {
+						warn!("Unsupported input token for the witnessed solana vault swap, omitting the swap and the swap account.");
+						None
+					},
+				},
+			))
+		})
+		.try_flatten()
+		.try_collect()
```

Comment:
Good point. Another one that should be handled when the changes are done like above. We should end up with no error @ramizhasan111

Reply:
We can see in the code that it will never return an ```Err```. The items that error are filtered out in the inner ```filter_map``` and the iterator returned should always have non-error items and so it should be ok.

That said, I think this can be optimized as I feel there is needless layering of iterators. for example, I am not sure what purpose ```map_ok``` is serving here.

Reply:
ignore the above comment. while the inner closure always returns Ok, we would still need to deal with errors in rpc calls (basically ignore queries that errored).

Reply:
see commit [`ac586b6` (#5355)](https://github.com/chainflip-io/chainflip-backend/pull/5355/commits/ac586b60eaae369d9d1eeb2c8203fb79d9d7ad77)